### DISCUSSION TITLE
Provide more "semantic" access to things currently from FrameworkUtil

### DISCUSSION THREAD

**Discussion by laeubi** [osgi/osgi#544] [Ideas]
[`FrameworkUtil`](https://github.com/osgi/osgi/blob/main/org.osgi.framework/src/org/osgi/framework/FrameworkUtil.java) has currently some static method to get a Bundle 
https://github.com/osgi/osgi/blob/743a80f61d551345bb6c8a7d4c21f3aaa861ef44/org.osgi.framework/src/org/osgi/framework/FrameworkUtil.java#L210
https://github.com/osgi/osgi/blob/743a80f61d551345bb6c8a7d4c21f3aaa861ef44/org.osgi.framework/src/org/osgi/framework/FrameworkUtil.java#L226
and a Filter
https://github.com/osgi/osgi/blob/743a80f61d551345bb6c8a7d4c21f3aaa861ef44/org.osgi.framework/src/org/osgi/framework/FrameworkUtil.java#L84

But from code point of view its rather bad to rely on static util class:

1. One needs to know the name of the "util" and that it might offer something I currently try to archive
2. Method names must be long to describe what is actually done
3. many different concerns refer to the same class and might increase the size of method signatures

Instead of this, in recent java versions static interface methods have become quite popular as they are quite obvious to find, related to the objects return value and gives the reader a clear hint of the context, for example:

- https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#of-T-
- https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#of-T-
- https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#concat-java.util.stream.Stream-java.util.stream.Stream-
- https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/List.html#of()
- ...

I think it would be good to adapt this for OSGi as well to have:

- `Filter.of(String)`
- `Bundle.of(Class<?>)`
- `Bundle.of(ClassLoader<?>)`

these then can just delegate the the `FrameworkUtil` calls.

For sure this is completely "non-functional" and only adds some kind of syntactic sugar but would make code using these much more readable and more align to modern code styles.

**Comment by HannesWell** [comment]
I like this proposal.
> these then can just delegate the the FrameworkUtil calls.

It could even be considered to move the actual implementation to the `.of()` methods and let `FrameworkUtil` call that. That's how the JDK did it with `Path.of()` and `Paths.get()` (which was used before static interface methods where allowed).

**Comment by bjhargrave** [comment]
This does not add any real value to the spec. You can criticize the prior API name choices, but just replicating them under the current naming fashion does not add real value and only create confusion of why there are now 2+ ways to do a thing. So -1 from me.

**Reply by laeubi** [reply]
So the "value" of the spec is just to exist in its own " ivory tower"? I really don't get this habit of "no value" if there are requests from user obviously it would be valuable for them... I also gave several references to java where it was seen as valuable in contrast of e.g. just using the old static util classes (e.g. `Collections.XYZ` methods) ... so I'm a bit "confused" how easy OSGi users seem to be confused compared to the "regular" java developers ;-)

**Reply by HannesWell** [reply]
It is of course not ideal to have multiple ways of achieving the same result and one would not design it that way from scratch. But in a long living code like OSGi, I think multiple ways are ok as a result of evolution and that it is adapting new possibilities.
To recap the example from above with `Paths.get()` and `Path.of()` you have two identical ways in Java to create `Path` objects. There is not even that much difference in typing, nevertheless it was seen to add enough value to Java by the Java-devs to add `Path.of()` when it was possible.
Even in OSGi there are already multiple ways of achieving the same. For example you can probably do most of the things you do with DS with the low-level OSGi `BundleContext`, `ServiceReference`, `ServiceRegistration` and listeners. But with DS it is usually more convenient, save and faster to do, isn't it?
Of course this is 'syntactic sugar', but I would not see the existence of another less accessible and convenient way as a strong argument against it.

**Reply by mikerumpf** [reply]
I really don't get why good suggestions are "blocked" by claiming "it's not adding any value".  Moving OSGi-API closer to the "CleanCode"-paradigma is absolultely worth the effort. I am working in huge projects with OSGi. And everything that helps to create cleaner, more readable source code is welcomed here since it helps the next colleague who has to read & understand what others have written.

**Reply by tjwatson** [reply]
One -1 does not equate to "blocked".  Every contributor has the same ability to vote, with no-one having more power behind their vote.  Power is in the hands of the ones that do the work to make the specification happen.

There is more to this `FrameworkUtil` class than simply making it more pro-"CleanCode".  Frameworks are allowed to replace implementations of API.  Generally speaking there is very little implementation in the OSGi core framework APIs by design.  Most all APIs in core are interfaces with no implementation except for permission classes and this the unfortunate `FrameworkUtil` class.  Framework's often will replace the `FrameworkUtil` class with their own implementation.  If we disperse the actual implementation out to interfaces like `Bundle` and `Filter` then framework implementations may want/need to replace the implementations of the methods in these interfaces also.  That is no concern for the users of OSGi, but it is an added complication of the specification that we do need to deal with when defining the specification.

### FrameworkUtil.createFilter

I wish I could recall why we ever created this method at all.  The only core API that takes a `Filter` is `org.osgi.util.tracker.ServiceTracker.ServiceTracker(BundleContext, Filter, ServiceTrackerCustomizer<S, T>)`.  But this method already takes a `BundleContext` so the `BundleContext.createFilter` can be used there.  Besides that, the `ServiceTracker` only ever calls `toString` on the passed in `Filter` object which seems like a waste to even take a `Filter` instead of simply taking a `String` for the filter.  The current `ServiceTracker` that takes a `Filter` object requires double parsing of the `Filter` for no added functional value.

I do not think we need a third way to create a `Filter` object when I see little value to the current two other methods for creating `Filter` objects.

### FrameworkUtil.getBundle methods
What makes this more complicated are these methods and their relation to the `org.osgi.framework.connect.FrameworkUtilHelper`.  The helper can be used to provide implementation of the `getBundle` methods from connect implementations outside of the framework.  Moving the `getBundle` to someplace else would need to explain how they relate to `FrameworkUtilHelper`.  Not impossible to explain, but not something I would call "CleanCode" either.

**Reply by bjhargrave** [reply]
> I wish I could recall why we ever created this method at all.

Other specs use filter strings. So this method enables parsing them into `Filter` object without the need for the method to have been passed a `BundleContext`.


> Framework's often will replace the `FrameworkUtil` class with their own implementation.

Yes. `FrameworkUtil` is the choke point such that framework implementations only need to replace one class to customize a set of default implementations. Had `FrameworkUtil` existed in Release 1, we may not have ever defined the `BundleContext.createFilter` method.

**Reply by tjwatson** [reply]
> Other specs use filter strings. So this method enables parsing them into `Filter` object without the need for the method to have been passed a `BundleContext`.

But the implementation behind the other specification would have had to be inside an installed/started bundle, so they would have the `BundleContext` of their own to call `createFilter` under the covers.

**Reply by bjhargrave** [reply]
> But the implementation behind the other specification would have had to be inside an installed/started bundle, so they would have the `BundleContext` of their own to call `createFilter` under the covers.

Yes, but they would need to make arrangements to pass the `BundleContext` all the way down the call stack to the method which needed it (or stuff it in some well known static variable). This can be cumbersome to the code base. Especially since it could require methods in intermediate call stack frames to be OSGi-aware which may not always be possible if using 3rd party libraries.

**Reply by laeubi** [reply]
@tjwatson this clearly depends on the implementation, just assume I have a bundle that contains code to work with filters but d not register any services/components, I then require to either create an activator and use some ugly static access, use `FrameworkUtil` (!) to get my bundle and hope it is started (or mark it lazy-start), just to use the bundlecontext to create a filterobject that actually don't care about bundle start/stop as it is completely stateless...

**Reply by tjwatson** [reply]
I don't believe we added `FrameworkUtil.createFilter` for any of these reasons.  The only concrete reason I recall that we added this method was to make implementations of our various permission classes easier that require a `Filter` implementation under the covers.  I can agree that, if it exists then someone will find a reason to use it.

> to get my bundle and hope it is started (or mark it lazy-start)

On this point, that is a mistake in p2 that is proliferating into trying to make changes to the OSGi specification.  Any bundle installed into the framework should always be marked for activation according to its activation policy (lazy or eager).  If not, then I question why the bundle is even installed at all.

**Reply by laeubi** [reply]
> Any bundle installed into the framework should always be marked for activation according to its activation policy (lazy or eager). If not, then I question why the bundle is even installed at all.

I'm not sure I understand this, but if I install a bundle and then resolve it, I can use classes from it, but ti will not have a bundle-context to create filters, but other bundles (that are started) might still find it useful...

**Reply by tjwatson** [reply]
There have been so many issues and confusion over the decades around the lazy activation handling in p2 (and its predecessor Eclipse Update).  Bundles marked for lazy activation are magically just treated as always available to be in the `STARTING` state waiting for a class to load to activate all the while these lazy activated bundles have a valid BundleContext.  But bundles that do not use the lazy activation policy (which is most all others outside of the bundles that want to participate in the Eclipse platform) are not marked for activation unless some extra steps are taking to have a configuration unit associated with the bundle to have p2 actually start it.

This is not the norm for any other OSGi framewrok usage, in my experience.  Normally bundles that are installed are also marked for activation.  Not marking the bundle for activation would be the corner case.  Unfortunately p2 makes that the norm and makes it difficult for "normal" bundles to be eagerly activated like they would expect in most all other OSGI framework usage.
